File: HisabCore/Sources/HisabCore/PaytmPDFParser.swift
import Foundation
#if canImport(PDFKit)
import PDFKit
#endif

/// Text-layer parser for Paytm "UPI Statement" PDFs. Block shape:
///
///     02 Sep                      (no year — inferred from the header period)
///     7:33 PM
///     Paid to FirstClub           (may wrap; "Received from …" for credits)
///     UPI ID: …                    (optional)
///     UPI Ref No: 624536311139
///     Note: …  /  Tag: + "# …"     (optional)
///     HDFC Bank -                  (account column, wraps)
///     93
///     - Rs.453                     ("+ Rs.600" for credits)
///
/// PDFKit's reading order interleaves amounts inline on some pages and emits them
/// all after the blocks on others, so each page pairs its k-th block with its k-th
/// signed amount. Signed amounts before a page's first block (the page-1 summary)
/// are ignored. A page header can arrive glued to the first date
/// ("…Your Account 24 Aug").
public enum PaytmPDFText {
    private static var dateRegex: Regex<Substring> { /^\d{2} [A-Z][a-z]{2}$/ }
    private static var timeRegex: Regex<(Substring, Substring)> { /^\d{1,2}:\d{2}[\u{202F} ]?(AM|PM)$/ }
    private static var amountRegex: Regex<(Substring, Substring, Substring)> { /^([+-]) ?Rs\.([\d,]+(?:\.\d{1,2})?)$/ }

    public static func parse(pages: [String]) throws -> ParsedDocument {
        let allLines = pages.flatMap { $0.components(separatedBy: .newlines) }
            .map { $0.trimmingCharacters(in: .whitespaces) }
        guard let period = PaytmXLSXParser.period(in: allLines) else {
            throw ParseError.unrecognizedFormat
        }

        var blocks: [[String]] = []
        var amounts: [Int64] = []  // signed paise, in stream order

        for page in pages {
            var pageBlocks: [[String]] = []
            var pageAmounts: [Int64] = []
            var current: [String]?

            for rawLine in page.components(separatedBy: .newlines) {
                var line = rawLine.trimmingCharacters(in: .whitespaces)
                if line.isEmpty { continue }

                // Header glued to the first date of the page.
                if let range = line.range(of: "Your Account ") {
                    let suffix = String(line[range.upperBound...])
                    if suffix.wholeMatch(of: dateRegex) != nil { line = suffix }
                }

                if let match = line.wholeMatch(of: amountRegex) {
                    if current != nil || !pageBlocks.isEmpty {  // ignore pre-table summary amounts
                        let magnitude = Money.signedPaise(fromDecimalString: String(match.2)) ?? 0
                        pageAmounts.append(match.1 == "-" ? -magnitude : magnitude)
                    }
                    continue
                }
                if line.wholeMatch(of: dateRegex) != nil {
                    if let block = current { pageBlocks.append(block) }
                    current = [line]
                    continue
                }
                current?.append(line)
            }
            if let block = current { pageBlocks.append(block) }
            guard pageBlocks.count == pageAmounts.count else {
                throw ParseError.malformedRow(0, "page has \(pageBlocks.count) blocks but \(pageAmounts.count) amounts")
            }
            blocks += pageBlocks
            amounts += pageAmounts
        }
        guard !blocks.isEmpty else { throw ParseError.empty }

        let formatter = DateFormatter()
        formatter.calendar = YearMonth.istCalendar
        formatter.timeZone = YearMonth.istCalendar.timeZone
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "dd MMM yyyy h:mm a"

        let startYear = YearMonth(date: period.start).year
        let endYear = YearMonth(date: period.end).year
        var previousDate: Date?
        var transactions: [ParsedTransaction] = []

        for (block, signedPaise) in zip(blocks, amounts) {
            guard block.count >= 3, block[1].wholeMatch(of: timeRegex) != nil else {
                throw ParseError.malformedRow(0, block.joined(separator: " | "))
            }
            let time = block[1].replacingOccurrences(of: "\u{202F}", with: " ")

            // Year inference: the candidate that falls inside the statement period wins;
            // ties (period spanning the same month twice) resolve via the statement's
            // newest-first ordering.
            var date: Date?
            for year in Set([endYear, startYear]).sorted(by: >) {
                guard let candidate = formatter.date(from: "\(block[0]) \(year) \(time)"),
                      candidate >= period.start, candidate <= period.end else { continue }
                if let chosen = date {
                    if let previous = previousDate, candidate <= previous, chosen > previous {
                        date = candidate
                    }
                } else {
                    date = candidate
                }
            }
            guard let date else {
                throw ParseError.malformedRow(0, block[0])
            }
            previousDate = date

            let fieldPrefixes = ["UPI ID:", "UPI Ref No:", "Note:", "Tag:", "#"]
            var detailLines: [String] = []
            for line in block.dropFirst(2) {
                if fieldPrefixes.contains(where: { line.hasPrefix($0) }) { break }
                detailLines.append(line)
            }
            guard let detail = detailLines.first else {
                throw ParseError.malformedRow(0, block.joined(separator: " | "))
            }
            var counterparty = detailLines.joined(separator: " ")
            for prefix in ["Paid to ", "Received from ", "Money sent to "] where detail.hasPrefix(prefix) {
                counterparty = String(counterparty.dropFirst(prefix.count))
                break
            }
            let reference = block.first { $0.hasPrefix("UPI Ref No:") }
                .map { $0.dropFirst("UPI Ref No:".count).trimmingCharacters(in: .whitespaces) }

            transactions.append(ParsedTransaction(
                date: date,
                amountPaise: abs(signedPaise),
                direction: signedPaise < 0 ? .debit : .credit,
                counterparty: counterparty,
                reference: reference,
                narration: block.dropFirst(2).joined(separator: " / ")))
        }

        return ParsedDocument(source: .paytm, declaredPeriod: period, transactions: transactions)
    }
}

#if canImport(PDFKit)
public struct PaytmPDFParser: StatementParser {
    public let source = Source.paytm

    public init() {}

    public func canParse(data: Data, filename: String) -> Bool {
        guard data.starts(with: Array("%PDF".utf8)) else { return false }
        guard let doc = PDFDocument(data: data) else { return false }
        if doc.isLocked {
            return filename.lowercased().contains("paytm")
        }
        return doc.page(at: 0)?.string?.contains("Paytm Statement for") ?? false
    }

    public func parse(data: Data, password: String?) throws -> ParsedDocument {
        guard let doc = PDFDocument(data: data) else { throw ParseError.unrecognizedFormat }
        if doc.isLocked {
            guard let password, doc.unlock(withPassword: password) else {
                throw ParseError.passwordRequired
            }
        }
        let pages = (0..<doc.pageCount).compactMap { doc.page(at: $0)?.string }
        return try PaytmPDFText.parse(pages: pages)
    }
}
#endif

File: hisab_flutter/packages/hisab_pdf/lib/src/paytm_pdf_parser.dart
/// Paytm "UPI Statement" PDFs over Syncfusion's line stream, which merges
/// each row into one visual line ending in the signed amount:
///
///     02 Sep Paid to FirstClub Note: UPIIntent HDFC Bank - - Rs.453
///     7:33 PM
///     UPI ID: firstclub-13003734.payu@indus Tag: 93
///     UPI Ref No: 624536311139 # Groceries
///
/// The date has no year — inferred from the statement period, matching the
/// Swift parser's newest-first tie-breaking.
library;

import 'package:hisab_core/hisab_core.dart';

import 'pdf_text.dart';

class PaytmPdfParser implements StatementParser {
  @override
  Source get source => Source.paytm;

  const PaytmPdfParser();

  static final _rowRegex = RegExp(
      r'^(\d{2}) ([A-Z][a-z]{2}) (.+?) ([+-]) ?Rs\.([\d,]+(?:\.\d{1,2})?)$');
  static final _timeRegex = RegExp(r'^(\d{1,2}):(\d{2})\s?(AM|PM)$');
  static final _refRegex = RegExp(r'UPI Ref No: (\d+)');
  static final _periodRegex =
      RegExp(r"(\d{1,2}) ([A-Z]{3})'(\d{2}) - (\d{1,2}) ([A-Z]{3})'(\d{2})");

  static const _months = {
    'Jan': 1, 'Feb': 2, 'Mar': 3, 'Apr': 4, 'May': 5, 'Jun': 6,
    'Jul': 7, 'Aug': 8, 'Sep': 9, 'Oct': 10, 'Nov': 11, 'Dec': 12,
  };
  static const _upperMonths = {
    'JAN': 1, 'FEB': 2, 'MAR': 3, 'APR': 4, 'MAY': 5, 'JUN': 6,
    'JUL': 7, 'AUG': 8, 'SEP': 9, 'OCT': 10, 'NOV': 11, 'DEC': 12,
  };

  @override
  bool canParse(List<int> data, String filename) {
    if (PdfText.isLocked(data, null)) {
      return filename.toLowerCase().contains('paytm');
    }
    final pages = PdfText.pageTexts(data, null);
    if (pages == null || pages.isEmpty) return false;
    return pages.first.contains('Paytm Statement for');
  }

  @override
  ParsedDocument parse(List<int> data, {String? password}) {
    List<String>? pages;
    try {
      pages = PdfText.pageTexts(data, password);
    } catch (_) {
      throw const PasswordRequiredException();
    }
    if (pages == null) throw const UnrecognizedFormatException();
    final lines = pages
        .expand((p) => p.split('\n'))
        .map((l) => l.trim())
        .where((l) => l.isNotEmpty)
        .toList();

    DatePeriod? period;
    for (final line in lines) {
      final match = _periodRegex.firstMatch(line);
      if (match == null) continue;
      final sm = _upperMonths[match.group(2)];
      final em = _upperMonths[match.group(5)];
      if (sm == null || em == null) continue;
      period = DatePeriod(
        DateTime.utc(2000 + int.parse(match.group(3)!), sm,
                int.parse(match.group(1)!), 0)
            .subtract(istOffset),
        DateTime.utc(2000 + int.parse(match.group(6)!), em,
                int.parse(match.group(4)!), 23, 59)
            .subtract(istOffset),
      );
      break;
    }
    if (period == null) throw const UnrecognizedFormatException();

    // Collect blocks: a row line plus its follower lines (time, refs).
    final blocks = <(RegExpMatch, List<String>)>[];
    for (final line in lines) {
      final rowMatch = _rowRegex.firstMatch(line);
      if (rowMatch != null) {
        // The summary section has no dd MMM rows, so any match is a real row.
        blocks.add((rowMatch, []));
        continue;
      }
      if (blocks.isNotEmpty) blocks.last.$2.add(line);
    }
    if (blocks.isEmpty) throw const EmptyDocumentException();

    final startYear = YearMonth.fromDate(period.start).year;
    final endYear = YearMonth.fromDate(period.end).year;
    final years = {endYear, startYear}.toList()..sort((a, b) => b.compareTo(a));

    DateTime? previousDate;
    final transactions = <ParsedTransaction>[];

    for (final (row, followers) in blocks) {
      final month = _months[row.group(2)];
      if (month == null) throw MalformedRowException(0, row.group(0)!);
      final day = int.parse(row.group(1)!);

      var hour = 12, minute = 0;
      for (final line in followers) {
        final time = _timeRegex.firstMatch(line);
        if (time != null) {
          hour = int.parse(time.group(1)!) % 12;
          if (time.group(3) == 'PM') hour += 12;
          minute = int.parse(time.group(2)!);
          break;
        }
      }

      // Year inference: the candidate inside the period wins; ties resolve
      // via the statement's newest-first ordering.
      DateTime? date;
      for (final year in years) {
        final candidate =
            DateTime.utc(year, month, day, hour, minute).subtract(istOffset);
        if (candidate.isBefore(period.start) || candidate.isAfter(period.end)) {
          continue;
        }
        if (date == null) {
          date = candidate;
        } else if (previousDate != null &&
            !candidate.isAfter(previousDate) &&
            date.isAfter(previousDate)) {
          date = candidate;
        }
      }
      if (date == null) throw MalformedRowException(0, row.group(0)!);
      previousDate = date;

      final magnitude = Money.signedPaise(row.group(5)!);
      if (magnitude == null || magnitude <= 0) {
        throw MalformedRowException(0, row.group(0)!);
      }
      final direction =
          row.group(4) == '-' ? Direction.debit : Direction.credit;

      String? reference;
      for (final line in followers) {
        final ref = _refRegex.firstMatch(line);
        if (ref != null) {
          reference = ref.group(1);
          break;
        }
      }

      var counterparty = row.group(3)!;
      for (final prefix in ['Paid to ', 'Received from ', 'Money sent to ']) {
        if (counterparty.startsWith(prefix)) {
          counterparty = counterparty.substring(prefix.length);
          break;
        }
      }
      for (final cut in [' Note:', ' Tag:', ' HDFC Bank', ' IDFC FIRST']) {
        final at = counterparty.indexOf(cut);
        if (at > 0) counterparty = counterparty.substring(0, at);
      }

      transactions.add(ParsedTransaction(
        date: date,
        amountPaise: magnitude,
        direction: direction,
        counterparty: counterparty.trim(),
        reference: reference,
        narration: '${row.group(3)} / ${followers.join(' / ')}',
      ));
    }

    return ParsedDocument(
        source: Source.paytm, declaredPeriod: period, transactions: transactions);
  }
}
